parsers/shared/src/main/scala/sigmastate/lang/ContractDocParser.scala

```diff
@@ -0,0 +1,124 @@
+ */
+private object DocumentationToken {
+  def apply(kind: Kind): DocumentationToken = new DocumentationToken(kind, None, None)
+
+  def apply(kind: Kind, body: String): DocumentationToken = new DocumentationToken(kind, None, Option(body))
+
+  def apply(kind: TagKind, name: String, body: String): DocumentationToken =
+    new DocumentationToken(kind, Option(name), Option(body))
+
+  /**
+   * Represents a documentation remark.
+   */
+  sealed abstract class Kind
+
+  /**
+   * Documents an untagged doc description, this is simply a line of text anywhere in the docstring.
+   */
+  case object Description extends Kind
+
+  /**
+   * Represents an empty doc line.
+   */
+  case object EmptyLine extends Kind
+
+  /**
+   * Represents a line starting with a tag (@) but is not supported.
+   */
+  case object UnsupportedTag extends Kind
+
+  /**
+   * Represents a labeled documentation remark.
+   */
+  sealed abstract class TagKind(val label: String) extends Kind
+
+  /**
+   * Documents a specific value parameter of the contract template.
+   */
+  case object Param extends TagKind("@param")
+
+  /**
+   * Documents the return value of the contract template.
+   */
+  case object Return extends TagKind("@returns")
+}
+
+/**
+ * Holds values extracted from a `@param` tag line in docstrings.
+ * @param name Name of the parameter.
+ * @param description Description of the parameter.
+ */
+case class ParameterDoc(name: String, description: String)
+
+/**
+ * Contract template documentation extracted from the preceding docstring.
+ * @param description Top level contract template description.
+ * @param params Contract template parameters as defined in the docstring.
+ */
+case class ContractDoc(description: String, params: Seq[ParameterDoc])
+
+object ContractDoc {
+  def apply(tokens: Seq[DocumentationToken]): ContractDoc = {
+    val nonEmptyTokens = tokens.dropWhile(_.kind == DocumentationToken.EmptyLine)
+
+    val (description, paramTokens) = nonEmptyTokens.span(_.kind == DocumentationToken.Description)
+    val descriptionText = description.flatMap(_.body).mkString(" ")
+
+    def extractParamDocs(tokens: Seq[DocumentationToken]): Seq[ParameterDoc] = {
+      tokens match {
+        case DocumentationToken(kind: DocumentationToken.TagKind, Some(name), Some(body)) +: tail if kind == DocumentationToken.Param =>
+          // grab the succeeding description tokens if there are any, this is multiline description of params
+          val (descTokens, remainingTokens) = tail.span(_.kind == DocumentationToken.Description)
+          val fullDescription = (body +: descTokens.flatMap(_.body)).mkString(" ")
+          ParameterDoc(name, fullDescription) +: extractParamDocs(remainingTokens)
+        case _ +: tail => extractParamDocs(tail)
+        case Seq() => Seq()
+      }
+    }
+
+    ContractDoc(descriptionText, extractParamDocs(paramTokens))
+  }
+}
+
+/** Parser for docstrings of contract templates. */
+object ContractDocParser {
```

Comment:
I don't think we need a separate object. Does it make sense as an independent parser? All the parsers below can just be moved to ContractParser.